toxcore/friend_requests.c

```diff
@@ -28,21 +28,278 @@
 #include "friend_requests.h"
 #include "util.h"
 
+#include <assert.h>
+
+
+// Nospam functions
+void fr_nospam_init(No_Spam *ns)
+{
+    memset(ns, 0, sizeof(No_Spam));
+}
+
+bool fr_nospam_set(No_Spam *ns, uint32_t value, const uint8_t *description, size_t description_length)
+{
+
+    if (description_length > MAX_NOSPAM_DESCRIPTION_LENGTH)
+        return false;
+
+    ns->nospam = value;
+
+    if (value == 0) {
+        ns->description_length = 0;
+        return true;
+    }
+
+    if (description_length == 0)
+        return true;
+    if (description == 0) {
+        ns->description_length = 0;
+        return true;
+    }
+
+    ns->description_length = description_length;
+    memcpy(ns->description, description, ns->description_length);
+
+    return true;
+}
+
+uint32_t fr_nospam_value(const No_Spam *ns)
+{
+    return ns->nospam;
+}
+
+bool fr_nospam_valid(const No_Spam *ns)
+{
+    return (fr_nospam_value(ns) != NOSPAM_SPAM);
+}
+
+size_t fr_nospam_description_length(const No_Spam *ns)
+{
+    return ns->description_length;
+}
+
+void fr_nospam_description(const No_Spam *ns, uint8_t *buffer)
+{
+    memcpy(buffer, ns->description, ns->description_length);
+}
+
+size_t fr_nospam_saved_size(const No_Spam *ns)
+{
+    return sizeof(No_Spam);
+}
+
+void fr_nospam_save(const No_Spam *ns, uint8_t *buffer)
+{
+    memcpy(buffer, ns, sizeof(No_Spam));
```

Comment:
6. You don't memcpy a struct, it's not portable at all. You should memcpy each member of it separably.

7. You also should convert everything to the right endianness when saving and convert it back when loading, otherwise it won't be portable across systems with different byte order.